mllib/src/main/scala/org/apache/spark/ml/feature/TargetEncoder.scala

```diff
@@ -0,0 +1,447 @@
+                  }
+                  if (value < 0.0 || value != value.toInt) throw new SparkException(
+                    s"Values from column ${inputFeatures(feature)} must be indices, " +
+                      s"but got $value.")
+                  else Some(value) // non-null category
+                }
+              }
+              val (class_count, class_stat) = agg(feature).getOrElse(category, (0.0, 0.0))
+              val (global_count, global_stat) =
+                agg(feature).getOrElse(TargetEncoder.UNSEEN_CATEGORY, (0.0, 0.0))
+              $(targetType) match {
+                case TargetEncoder.TARGET_BINARY => // counting
+                  if (label == 1.0) {
+                    // positive => increment both counters for current & unseen categories
+                    agg(feature) +
+                      (category -> (1 + class_count, 1 + class_stat)) +
+                      (TargetEncoder.UNSEEN_CATEGORY -> (1 + global_count, 1 + global_stat))
+                  } else if (label == 0.0) {
+                    // negative => increment only global counter for current & unseen categories
+                    agg(feature) +
+                      (category -> (1 + class_count, class_stat)) +
+                      (TargetEncoder.UNSEEN_CATEGORY -> (1 + global_count, global_stat))
+                  } else throw new SparkException(
+                    s"Values from column ${getLabelCol} must be binary (0,1) but got $label.")
+                case TargetEncoder.TARGET_CONTINUOUS => // incremental mean
+                  // increment counter and iterate on mean for current & unseen categories
+                  agg(feature) +
+                    (category -> (1 + class_count,
+                      class_stat + ((label - class_stat) / (1 + class_count)))) +
+                    (TargetEncoder.UNSEEN_CATEGORY -> (1 + global_count,
+                      global_stat + ((label - global_stat) / (1 + global_count))))
+              }
+            }
+          }.toArray
+        } else agg,  // ignore null-labeled observations
+        (agg1, agg2) => inputFeatures.indices.map {
+          feature => {
+            val categories = agg1(feature).keySet ++ agg2(feature).keySet
+            categories.map(category =>
+              category -> {
+                val (counter1, stat1) = agg1(feature).getOrElse(category, (0.0, 0.0))
+                val (counter2, stat2) = agg2(feature).getOrElse(category, (0.0, 0.0))
+                $(targetType) match {
+                  case TargetEncoder.TARGET_BINARY => (counter1 + counter2, stat1 + stat2)
+                  case TargetEncoder.TARGET_CONTINUOUS => (counter1 + counter2,
+                    ((counter1 * stat1) + (counter2 * stat2)) / (counter1 + counter2))
+                }
+              }).toMap
+          }
+        }.toArray)
+
+    // encodings: Map[feature, Map[Some(category), encoding]]
+    val encodings: Map[String, Map[Option[Double], Double]] =
```

Comment:
I feel this is computation is not very complex and maybe implemented with sql functions.
but I am also fine to start with a RDD implementation.